### PR TITLE
[WIP] First version to add mutable collections

### PR DESCRIPTION
**Pull request by odersky**
This is currently WIP.

### CODE REVIEW

File: src/main/scala/strawman/collection/mutable/Seq.scala [hunk 1]
```diff
@@ -0,0 +1,76 @@
+package strawman
+package collection
+package mutable
+
+trait Seq[A] extends strawman.collection.Seq[A] with Growable[A] {
```
Comment (Ichoran) [inline src/main/scala/strawman/collection/mutable/Seq.scala]: This means that `Array` cannot be wrapped as a `Seq`.  Is this intentional?

File: src/main/scala/strawman/collection/mutable/Seq.scala [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+  // +=, ++=, clear inherited from Growable
+  def +=:(elem: A): this.type = { insert(0, elem); this }
+  def +=:(elem1: A, elem2: A, elems: A*): this.type = elem1 +=: elem2 +=: elems.toStrawman ++=: this
+  def ++=:(elems: IterableOnce[A]): this.type = { insertAll(0, elems); this }
+
+  def dropInPlace(n: Int): this.type = { remove(0, n); this }
+  def dropRightInPlace(n: Int): this.type = { remove(length - n, n); this }
+  def takeInPlace(n: Int): this.type = { remove(n, length); this }
+  def takeRightInPlace(n: Int): this.type = { remove(0, length - n); this }
+  def sliceInPlace(start: Int, end: Int): this.type = takeInPlace(end).dropInPlace(start)
```
Comment (Ichoran) [inline src/main/scala/strawman/collection/mutable/Seq.scala]: As a strawman this is fine, but obviously it will need to be overridden for collections that can't take/drop efficiently (e.g. because they don't have a last element index which makes `takeInPlace` O(1)).

**Comment by Ichoran** [comment]
This looks reasonable.  I think we might need more fine-grained traits to distinguish between fixed-length collections (where the contents can be altered) and growable things, but that doesn't undermine the strategy.

File: src/main/scala/strawman/collection/mutable/Seq.scala (hunk 2 above)
Comment (odersky) [inline src/main/scala/strawman/collection/mutable/Seq.scala]: My assumption was that most mutable sequences have indexed access. The only exception I know of is LIstBuffer. And, yes, `sliceInPlace` should be overridden for this one.

File: src/main/scala/strawman/collection/mutable/Seq.scala (hunk 1 above)
Comment (odersky) [inline src/main/scala/strawman/collection/mutable/Seq.scala]: No, not intentional. I am not sure yet what the right scehem should be. It seems natural that mutable sequences should not just have `update`, but also `insert`, `remove`, `+=`. But then what to do about `WrappedArray`? One possibility is to also have these things for `WrappedArray` and accept that e.g. `+=` has linear cost. That's what current collections seem to do. Another possibility would be to back `WrappedArray` by a `ResizableArray`.

File: src/main/scala/strawman/collection/mutable/Seq.scala (hunk 2 above)
Comment (Ichoran) [inline src/main/scala/strawman/collection/mutable/Seq.scala]: `UnrolledBuffer` is another one that we have, and a `SortedSeq` backed by a tree would be another.  It might fit better in `IndexedOptimizedSeq`, since there presumably the exceptions would be vanishingly rare (and most everything else would be an `IndexedOptimizedSeq` anyway).

File: src/main/scala/strawman/collection/mutable/Seq.scala (hunk 1 above)
Comment (Ichoran) [inline src/main/scala/strawman/collection/mutable/Seq.scala]: The problem with having `+=` and other Growable operations produce a new collection on `WrappedArray` is that your `inPlace` operations no longer operate on the collection you wrapped, and it's not at all obvious that this is what happened, since you might go back to the original array and expect to find the changes there.  One could also envision a mutable proxied Seq backed by whatever was useful (not just `Array`)--the equivalent of a lens for a set of mutable values.  We needn't necessarily support these sorts of things, but I think the cost would be relatively low, just like having a distinction between `Seq` and `IndexedOptimizedSeq` is (hopefully) low.

Comment (odersky) [inline src/main/scala/strawman/collection/mutable/Seq.scala]: Yes, good point. Current collections ignore this issue. Do people have proposals how to solve this? It seems there's no nice option.

File: src/main/scala/strawman/collection/mutable/Iterator.scala [hunk 3]
```diff
@@ -4,9 +4,10 @@ import scala.{Boolean, Int, Unit, Nothing, NoSuchElementException}
 import strawman.collection.{IndexedView, IterableOnce}
 
 /** A core Iterator class */
-trait Iterator[+A] { self =>
+trait Iterator[+A] extends IterableOnce[A] { self =>
```
Comment (julienrf) [inline src/main/scala/strawman/collection/mutable/Iterator.scala]: Is this used somewhere? I think it is better to avoid cycles in the hierarchy. Here, the problem is that if someone wants to understand the implementation of the collections it is hard to find a starting point. It is hard to find a first abstraction layer on top of which other abstraction layers are built. If we have a look at any collection and follow its ancestors, we always end up on `IterableOnce`. This one has an `iterator(): Iterator[A]` member, if we have a look at the `Iterator` type, we see that it also exends `IterableOnce`, etc.

BTW, I’m not sure this is related but the new design leads to stack overflowing the scala compiler extremely easily, making it super unproductive to diagnose type errors… This might be due to the cycles in the class hierarchy (there is also one with `View`).

Comment (odersky) [inline src/main/scala/strawman/collection/mutable/Iterator.scala]: It's used implicitly in that lots of operations use `IterableOnce` parameters (analogous to `TraversableOnce in the current collections), and the idea is that both an Iterator and an Iterable are valid arguments for this.

File: src/main/scala/strawman/collection/mutable/Seq.scala (hunk 1 above)
Comment (Ichoran) [inline src/main/scala/strawman/collection/mutable/Seq.scala]: Is there something wrong with the "obvious" solution of having the `Growable` trait implemented on only a subset of collections?  So we'd have `GrowableSeq` that extends both `Growable` and `Seq`, or alternatively `UpdatingSeq` which only allows update, and have `Seq` be `UpdatingSeq with Growable`?

Comment (odersky) [inline src/main/scala/strawman/collection/mutable/Seq.scala]: This is what's wrong:

    scala> collection.mutable.Seq(1, 2, 3)
    res1: scala.collection.mutable.Seq[Int] = ArrayBuffer(1, 2, 3)

So far, even the default `Seq` naturally supports `+=`. I don't think we can take that away. So the less drastic change would probably be to have a `FixedLengthSeq` superclass of `Seq` and have `WrappedArray` inherit from that instead of `Seq`.

Comment (Ichoran) [inline src/main/scala/strawman/collection/mutable/Seq.scala]: That is true, but on the other hand

```scala
scala> val s = collection.mutable.Seq(1)
s: scala.collection.mutable.Seq[Int] = ArrayBuffer(1)

scala> s += 2
<console>:12: error: value += is not a member of scala.collection.mutable.Seq[Int]
       s += 2
         ^
```

so people who are going by the API and not by pattern-matching can't rely on it even now.  And that `Seq` might very well be a `WrappedArray` that doesn't have the operation now.

So I think there's a decent argument for leaving it alone, too.  But having `+=` as a core functionality (and one that was usually lurking there one pattern match away, even if you couldn't count on it) is appealing too, so `FixedLengthSeq` (or maybe `SizedSeq` so it's not too wordy?) is also reasonable.

Comment (odersky) [inline src/main/scala/strawman/collection/mutable/Seq.scala]: Ah, cool. I completely misremembered what `mutable.Seq` was so far. I thought it had `+=` as well as `update`. So, yes, indeed, we can have `GrowableSeq` underneath `Seq`.

File: src/main/scala/strawman/collection/mutable/Iterator.scala (hunk 3 above)
Comment (julienrf) [inline src/main/scala/strawman/collection/mutable/Iterator.scala]: I’m still afraid that this could lead to very surprising behaviors. Consider for instance the following program:

~~~ scala
def f(xs: TraversableOnce[Int]): Unit = {
  println(xs.iterator().next())
  println(xs.iterator().next())
}
~~~

We basically create a new iterator and print its first element, twice in a row. If I run this program I would expect to see the same value printed twice. However, if I pass an `Iterator[Int]` as a parameter to `f`, it will print the first element and then the second element. That’s because `iterator()`’s implementation in `Iterator` returns `this` instead of a fresh iterator.

Comment (odersky) [inline src/main/scala/strawman/collection/mutable/Iterator.scala]: But that's why it says `IterableOnce`! It means you can take one `iterator()`, but not several ones. Note that `TraversableOnce` was not part of 2.8 collections. It got added later by @paulp, because we experienced a lot of pain regarding the divergence of `Iterator` and `Iterable`.

File: src/main/scala/strawman/collection/mutable/ArrayBuffer.scala [hunk 4]
```diff
@@ -78,16 +121,46 @@ object ArrayBuffer extends IterableFactory[ArrayBuffer] {
...
+
+  /** Remove elements of this array at indices after `sz`.
+   */
+  def nullElems(array: Array[AnyRef], start: Int, end: Int): Unit = {
+    var i = start
+    while (i < end) {
+      array(i) = null
+      i += 1
+    }
+  }
```
Comment (Ichoran) [inline src/main/scala/strawman/collection/mutable/ArrayBuffer.scala]: Except for very short arrays, this would be faster as `java.util.Arrays.fill(array, start, end, null)`

**Comment by Ichoran** [comment]
Looks good to me!

File: src/main/scala/strawman/collection/mutable/ArrayBuffer.scala [hunk 5]
```diff
@@ -31,45 +43,76 @@ class ArrayBuffer[A] private (initElems: Array[AnyRef], initLength: Int)
 
   protected[this] def newBuilder = new ArrayBuffer[A]
 
+  def clear() =
```
Comment (julienrf) [inline src/main/scala/strawman/collection/mutable/ArrayBuffer.scala]: Minor: you should add an explicit return type

File: src/main/scala/strawman/collection/mutable/ArrayBuffer.scala [hunk 6]
```diff
@@ -31,45 +43,76 @@ class ArrayBuffer[A] private (initElems: Array[AnyRef], initLength: Int)
...
-  /** Overridden to use array copying for efficiency where possible. */
-  override def ++[B >: A](xs: IterableOnce[B]): ArrayBuffer[B] = xs match {
-    case xs: ArrayBuffer[B] =>
-      val elems = new Array[AnyRef](length + xs.length)
-      Array.copy(this.elems, this.start, elems, 0, this.length)
-      Array.copy(xs.elems, xs.start, elems, this.length, xs.length)
-      new ArrayBuffer(elems, elems.length)
-    case _ => super.++(xs)
+  def insertAll(idx: Int, elems: IterableOnce[A]): Unit = {
+    if (idx < 0 || idx > end) throw new IndexOutOfBoundsException
```
Comment (julienrf) [inline src/main/scala/strawman/collection/mutable/ArrayBuffer.scala]: What about introducing a private method `ensureWithinBounds(i: Int): Unit`?

File: src/main/scala/strawman/collection/mutable/Growable.scala [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+import scala.annotation.tailrec
+import strawman.collection.{toOldSeq, toNewSeq}
+
+/** This trait forms part of collections that can be augmented
+ *  using a `+=` operator and that can be cleared of all elements using
+ *  a `clear` method.
+ *
+ *  @author   Martin Odersky
+ *  @version 2.8
+ *  @since   2.8
```
Comment (julienrf) [inline src/main/scala/strawman/collection/mutable/Growable.scala]: 2.13?

File: src/main/scala/strawman/collection/mutable/Builder.scala [hunk 8]
```diff
@@ -1,27 +1,27 @@
...
-trait Builder[-A, +To] { self =>
+trait Builder[-A, +To] extends Growable[A] { self =>
 
   /** Append an element */
   def +=(x: A): this.type
 
+  /** Clears the contents of this builder.
+   *  After execution of this method the builder will contain no elements.
+   */
+  def clear(): Unit
```
Comment (julienrf) [inline src/main/scala/strawman/collection/mutable/Builder.scala]: Note that we already inherit `clear` from `Growable`. I’m curious if there is a reason to re-define it here (maybe just to update the scaladoc?).

Comment (odersky) [inline src/main/scala/strawman/collection/mutable/Builder.scala]: Probably. I copy-pasted that one from existing sources.

File: src/main/scala/strawman/collection/mutable/Growable.scala (hunk 7 above)
Comment (odersky) [inline src/main/scala/strawman/collection/mutable/Growable.scala]: Yes, copy-paste fall-out.

File: src/main/scala/strawman/collection/mutable/ArrayBuffer.scala (hunk 6 above)
Comment (odersky) [inline src/main/scala/strawman/collection/mutable/ArrayBuffer.scala]: Yes, maybe. But it's not universal. We sometimes need two values to compare.

**Comment by odersky** [comment]
I'll work on this and #18 this Sunday to address reviewers comments. If we add maps and sets it would be good to do this on top of the fleshed out structure of #18.

File: src/main/scala/strawman/collection/mutable/Builder.scala (hunk 8 above)
Comment (odersky) [inline src/main/scala/strawman/collection/mutable/Builder.scala]: Note that the same holds for `+=`. It's also added in `Builder` for documentation.

**Comment by odersky** [comment]
Superseded by #18
